### PR TITLE
Animate: refactor to TypeScript

### PR DESCRIPTION
**Pull request by ciampo**
## What?

Refactor the `Animate` component to TypeScript

## Why?

Part of #35744

## How?

- Rewrite JSDoc types in TypeScript, apply to `Animate` component
- Rewrite Storybook examples to use controls and TS-generated docs

Reviewing commit-by-commit may help.

## Testing Instructions

- Project builds as before (type checks were already in place via previous JSDocs)
- Storybook examples work like on `trunk`

### CODE REVIEW

File: packages/components/src/animate/index.tsx
Comment (ciampo) [inline packages/components/src/animate/index.tsx:1]: Note that there shouldn't be runtime changes in this file

File: packages/components/src/animate/types.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+export type GetAnimateOptions =
+	| AppearOptions
+	| SlideInOptions
+	| LoadingOptions
+	| NoAnimationOptions;
+
+// Create a new type that and distributes the `Pick` operator separately to
+// every individual type of a union, thus preserving that same union.
+type DistributiveTypeAndOptions< T extends { type?: any } > = T extends any
```
Comment (ciampo) [inline packages/components/src/animate/types.ts:26]: @chad1008 does this look familiar?

File: packages/components/src/animate/stories/index.tsx
Comment (ciampo) [inline packages/components/src/animate/stories/index.tsx:1]: I rewrote this file to be more verbose, but in a way that allows us to get rid of the `Appear` helper and to use the `Story.args` syntax which works well with Storybook controls.

In short, I opted for more verbosity in exchange of more straightforward code.

File: packages/components/src/animate/index.tsx [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+ * 	</Animate>
+ * );
+ * ```
+ */
+export function Animate( { type, options = {}, children }: AnimateProps ) {
+	return children( {
+		className: getAnimateClassName( {
+			type,
+			...options,
+		} as GetAnimateOptions ),
```
Comment (ciampo) [inline packages/components/src/animate/index.tsx:71]: This typecast is not ideal, but there wasn't an easy way around it. Given that the `Animate` component doesn't seem to be currently used in Gutenberg, I thought I wouldn't spend too long on it.

File: packages/components/src/animate/index.tsx [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+ * 		{ ( { className } ) => (
+ * 			<Notice className={ className } status="success">
+ * 				<p>Animation finished.</p>
+ * 			</Notice>
+ * 		) }
+ * 	</Animate>
+ * );
+ * ```
+ */
+export function Animate( { type, options = {}, children }: AnimateProps ) {
```
Comment (tyxla) [inline packages/components/src/animate/index.tsx:66]: It felt odd that we're allowing `Animate` to not have any `type`, but I guess that this is by design. After all, if we want to not animate, we would just omit this component from the tree.

Anyway, this was like that before, and I've confirmed that the distributive conditional type support that use case correctly.

File: packages/components/src/animate/index.tsx (hunk 2 above)
Comment (tyxla) [reply packages/components/src/animate/index.tsx:71]: Fair enough 👍

**Review by tyxla — APPROVED** [review]
Looking and working great to me 🚀

File: packages/components/src/animate/index.tsx (hunk 3 above)
Comment (ciampo) [reply packages/components/src/animate/index.tsx:66]: It also felt odd to me, but then I saw that the Default Storybook example explicitly mentions that the lack of `type` prop means no animation 🤷

**Pull request merged**